AGENTS.md

```diff
@@ -0,0 +1,271 @@
+5. Build Pelican from source inside the container
+
+### Pull Request Guidelines
+
+- Address a single concern with minimal changes
+- Add tests for new functionality or bug fixes
+- Update documentation in `docs/` folder if needed
+- Follow the ["fork-and-pull" Git workflow](https://github.com/susam/gitpr)
+- At least one core contributor must review and approve changes
+- Run linters and tests before submitting
+
+### Configuration
+
+- Server configuration: YAML files (typically `pelican.yaml`)
+- Client configuration: Can use federation discovery URLs or explicit configuration
+- Parameters documented in `docs/parameters.yaml`
+
+## Key Features and Patterns
+
+### Object Transfer
+
+- Client commands: `pelican object get`, `pelican object put`, `pelican object copy`
+- Support for multiple protocols (HTTP, XRootD)
+- Checksum verification (MD5, SHA1, CRC32)
+- Resume capabilities for interrupted transfers
+
+### Federation Model
+
+- **Director**: Routes client requests to appropriate origins/caches
+- **Registry**: Manages namespace registrations
+- **Origins**: Serve data from storage backends
+- **Caches**: Provide distributed caching layer
+
+### Authentication
+
+- Token-based authentication with support for various issuers
+- OAuth integration via CILogon
+- Support for WLCG tokens
+
+### Monitoring
+
+- Prometheus metrics integration
+- Health checks for all services
+- Performance monitoring through web UI
+
+## API Documentation
+
+API endpoints are documented using OpenAPI V2.0. The specification is generated from code annotations.
+
+## Testing Philosophy
+
+- Unit tests for individual functions and components
+- Integration tests for service interactions
+- End-to-end tests in `e2e_fed_tests/`
+- Test utilities in `test_utils/` and `fed_test_utils/`
+
+## Security Considerations
+
+- TLS/HTTPS by default for all services
+- Token validation for authenticated operations
+- Regular dependency updates
+- CodeQL analysis in CI/CD pipeline
+
+## Documentation
+
+- User documentation: [docs.pelicanplatform.org](https://docs.pelicanplatform.org/)
+- API documentation: Available through web UI at `/api/docs`
+- Parameter reference: `docs/parameters.yaml`
+- HTTP headers: `docs/pelican-http-headers.md`
+- Error codes: `docs/error_codes.yaml`
+
+## Common Gotchas
+
+1. **XRootD Integration**: Origins and Caches require XRootD to be properly configured
+2. **MaxMind License**: Director requires a MaxMind license key for GeoIP lookups
+3. **OAuth Setup**: Registry requires CILogon OAuth credentials for web authentication
+4. **Port Conflicts**: Default ports may conflict with local services
+5. **TLS Certificates**: Development often uses `TLSSkipVerify: true` but production requires proper certificates
+
+## Useful Commands
+
+```bash
+# Run a local federation ("federation in a box")
+pelican serve --module director,registry,origin,cache
+
+# Serve an origin
+pelican origin serve -f https://director.example.com -v /tmp/stash/:/test
```

Comment:
When starting a server, it's always preferable to provide its configuration via a `pelican.yaml` file instead of through command line arguments. The exception to this rule is when you're testing command line arguments.